step/step.go

```diff
@@ -42,21 +67,227 @@ func New(logger log.Logger, inputParser stepconf.InputParser, commandFactory com
 func (step DockerBuildPushStep) Run() error {
 	var input Input
 	if err := step.inputParser.Parse(&input); err != nil {
-		return fmt.Errorf("failed to parse inputs: %w", err)
+		return fmt.Errorf("parse inputs: %w", err)
 	}
 	stepconf.Print(input)
 	step.logger.Println()
 
 	step.logger.EnableDebugLog(input.Verbose)
 
+	tagUsedInCacheKey := strings.Split(input.Tags, "\n")[0]
+
+	if input.UseBitriseCache {
+		if err := step.restoreCache(input, tagUsedInCacheKey); err != nil {
+			return fmt.Errorf("restore cache: %w", err)
+		}
+	}
+
+	if err := step.dockerBuild(input); err != nil {
+		return fmt.Errorf("build docker image: %w", err)
+	}
+
+	if input.UseBitriseCache {
+		if err := step.saveCache(input, tagUsedInCacheKey); err != nil {
+			return fmt.Errorf("save cache: %w", err)
+		}
+	}
+	return nil
+}
+
+func (step DockerBuildPushStep) restoreCache(input Input, imageName string) error {
+	step.logger.Infof("Restoring cache...")
+	saver := cache.NewRestorer(step.envRepo, step.logger, step.commandFactory)
+
+	var cacheKey = []string{
+		fmt.Sprintf(dockerCacheKeyTemplate, imageName),
+		fmt.Sprintf("docker-%s-{{ .OS }}-{{ .Arch }}-{{ .Branch }}", imageName),
+		fmt.Sprintf("docker-%s-{{ .OS }}-{{ .Arch }}", imageName),
+	}
+
+	return saver.Restore(cache.RestoreCacheInput{
+		StepId:  "restore-cache",
+		Verbose: input.Verbose,
+		Keys:    cacheKey,
+	})
+}
+
+func (step DockerBuildPushStep) saveCache(input Input, imageName string) error {
+	step.logger.Infof("Saving cache...")
+	saver := cache.NewSaver(step.envRepo, step.logger, step.pathProvider, step.pathModifier, step.pathChecker)
+
+	return saver.Save(cache.SaveCacheInput{
+		StepId:      "save-cache",
+		Verbose:     input.Verbose,
+		Key:         fmt.Sprintf(dockerCacheKeyTemplate, imageName),
+		Paths:       []string{dockerCacheFolder},
+		IsKeyUnique: false,
+	})
+}
+
+func (step DockerBuildPushStep) dockerBuild(input Input) error {
+	step.logger.Infof("Building docker image...")
+
+	if err := step.createCacheFolder(dockerCacheFolder); err != nil {
+		return fmt.Errorf("create cache folder: %w", err)
+	}
+	if err := step.createCacheFolder(dockerCacheFolderTemporary); err != nil {
+		return fmt.Errorf("create cache folder: %w", err)
+	}
+
+	buildkitContainer, err := step.initializeBuildkit(input)
+	if err != nil {
+		return fmt.Errorf("initialize buildkit: %w", err)
+	}
+	defer func() {
+		if err := step.destroyContainer(buildkitContainer); err != nil {
+			step.logger.Errorf("destroy buildx instance: %s", err)
+		}
+	}()
+
+	if err := step.build(input); err != nil {
+		return fmt.Errorf("build docker image: %w", err)
+	}
+
+	// To make sure the cache is not growing indefinitely,
+	// we remove the original cache folder and move the new one to its place
+	if err := step.moveCacheFolder(dockerCacheFolderTemporary, dockerCacheFolder); err != nil {
```

Comment:
Did you look into the `--filter` flag of `buildx prune`? https://docs.docker.com/engine/reference/commandline/buildx_prune/

Sounds like we could do something similar with recording the timestamp of starting the build, then clean up everything that was created before that timestamp? Some steps have this exact mechanism for dealing with files restored from cache vs. files produced during the build.

Reply:
We need the old cache during the build so pruneing before is not an option, pruning after using the time could be risky we could accidentally remove less or more. This solution has less variables and guaranteed to only cache exactly what we need.

Reply:
I'm okay with it, let's go this way first. We could iterate on it in the future